src/v/raft/replicate_batcher.cc

```diff
@@ -124,35 +145,42 @@ replicate_batcher::do_cache_with_backpressure(
      * When batch size exceed available semaphore units we just acquire all of
      * them to be able to continue.
      */
+    ssx::semaphore_units u;
+    if (timeout) {
+        u = co_await ss::get_units(
+          _max_batch_size_sem,
+          std::min(bytes, _max_batch_size),
+          ssx::semaphore::clock::now() + *timeout);
+    } else {
+        u = co_await ss::get_units(
+          _max_batch_size_sem, std::min(bytes, _max_batch_size));
+    }
 
-    return ss::get_units(_max_batch_size_sem, std::min(bytes, _max_batch_size))
-      .then(
-        [this, expected_term, batches = std::move(batches), consistency_lvl](
-          ssx::semaphore_units u) mutable {
-            size_t record_count = 0;
-            auto i = ss::make_lw_shared<item>();
-            for (auto& b : batches) {
-                record_count += b.record_count();
-                if (b.header().ctx.owner_shard == ss::this_shard_id()) {
```

Comment:
this commit is fairly difficult to grok. aside from the mixing of coroutinizing methods, it would even be useful if the commit message described how cancellation worked (e.g. are they really cancelled, or continue in the background? how is lifetime then managed, etc...?)

Reply:
I've split it into multiple commits